### PR TITLE
Calculate fees correctly based on zero/non-zero bytes

### PR DESCRIPTION
**Pull request by gakonst**
Fixes https://github.com/ethereum-optimism/optimism/issues/515. 

Does *not* introduce `nonCallDataL1GasOverhead` since we do not have a standard way to estimate / calculate it yet and seems to be a micro optimization.

### CODE REVIEW

File: l2geth/core/rollup_fee.go [hunk 1]
```diff
@@ -2,20 +2,33 @@ package core
...
 
 // CalculateFee calculates the fee that must be paid to the Rollup sequencer, taking into
 // account the cost of publishing data to L1.
-// Returns: (RollupBaseTxSize + len(data)) * dataPrice + executionPrice * gasUsed
+// Returns: (4 * zeroDataBytes + 16 * (nonZeroDataBytes + RollupBaseTxSize)) * dataPrice + executionPrice * gasUsed
 func CalculateRollupFee(data []byte, gasUsed uint64, dataPrice, executionPrice *big.Int) *big.Int {
-	dataLen := int64(RollupBaseTxSize + len(data))
+	var zeros uint64
+	for _, byt := range data {
+		if byt != 0 {
```
Comment (tynes) [inline l2geth/core/rollup_fee.go:20]: Should this be:
```
if byte == 0 {
    zeros++
}
```

File: l2geth/core/rollup_fee_test.go [hunk 2]
```diff
@@ -23,9 +23,21 @@ func TestCalculateRollupFee(t *testing.T) {
 			data := make([]byte, 0, tt.dataLen)
 			fee := CalculateRollupFee(data, tt.gasUsed, big.NewInt(tt.dataPrice), big.NewInt(tt.executionPrice))
 
-			dataFee := uint64((RollupBaseTxSize + len(data)) * int(tt.dataPrice))
+			var zeros uint64
```
Comment (tynes) [inline l2geth/core/rollup_fee_test.go:26]: This should go in a helper function if the same logic is duplicated in the code and the test

File: l2geth/core/rollup_fee.go [hunk 3]
```diff
@@ -2,20 +2,33 @@ package core
...
+			zeros++
+		}
+	}
+	ones := uint64(len(data)) - zeros
+
+	zerosCost := big.NewInt(int64(zeros * params.TxDataZeroGas))
```
Comment (tynes) [inline l2geth/core/rollup_fee.go:26]: To remove the intermediate type cast:
```golang
zerosCost := new(big.Int).SetUint64(zeros * params.TxDataZeroGas)
onesCost := new(big.Int).SetUint64((RollupBaseTxSize + ones) * params.TxDataNonZeroGasEIP2028)
```

**Review by tynes — CHANGES_REQUESTED** [review]
Left some comments and questions

**Comment by tynes** [comment]
Noting this for other reviewers: the correct spelling for the plural of "zero" is both "zeros" and "zeroes"

**Comment by tynes** [comment]
> Does not introduce nonCallDataL1GasOverhead since we do not have a standard way to estimate / calculate it yet and seems to be a micro optimization.

As long as we charge slightly more than the cost of submitting the batch + running the hardware, things should be fine. The profits will go towards funding public goods

File: integration-tests/test/native-eth.spec.ts [hunk 4]
```diff
@@ -41,17 +41,17 @@ describe('Native ETH Integration Tests', async () => {
   })
 
   describe('estimateGas', () => {
-    it('Should estimate gas for ETH transfer', async () => {
+    it.only('Should estimate gas for ETH transfer', async () => {
```
Comment (tynes) [inline integration-tests/test/native-eth.spec.ts:44]: I have made the mistake of leaving in a `only` as well, I feel like there is a syntax `$ mocha --only` that could prevent these types of problems from happening

Comment (gakonst) [reply integration-tests/test/native-eth.spec.ts]: My bad. Force-pushed should be gone now.

File: l2geth/core/rollup_fee.go [hunk 5]
```diff
@@ -2,21 +2,38 @@ package core
...
 // RollupBaseTxSize is the encoded rollup transaction's compressed size excluding
 // the variable length data.
 // Ref: https://github.com/ethereum-optimism/optimism/blob/91a9a3dcddf534ae1c906133b6d8e015a23c463b/packages/contracts/contracts/optimistic-ethereum/OVM/predeploys/OVM_SequencerEntrypoint.sol#L47
-const RollupBaseTxSize int = 96
+const RollupBaseTxSize uint64 = 96
 
 // CalculateFee calculates the fee that must be paid to the Rollup sequencer, taking into
 // account the cost of publishing data to L1.
-// Returns: (RollupBaseTxSize + len(data)) * dataPrice + executionPrice * gasUsed
+// Returns: (4 * zeroDataBytes + 16 * (nonZeroDataBytes + RollupBaseTxSize)) * dataPrice + executionPrice * gasUsed
```
Comment (tynes) [inline l2geth/core/rollup_fee.go:16]: Confirming the implementation:

`4 * zeroDataBytes` is `zeroesCost`
`16 * (nonZeroDataBytes + RollupBaseTxSize)` is `onesCost`
Their sum `(4 * zeroDataBytes + 16 * (nonZeroDataBytes + RollupBaseTxSize))` is `dataCost`
`(4 * zeroDataBytes + 16 * (nonZeroDataBytes + RollupBaseTxSize)) * dataPrice` is `dataFee`
`executionPrice * gasUsed` is `executionFee`
And finally,
`(4 * zeroDataBytes + 16 * (nonZeroDataBytes + RollupBaseTxSize)) * dataPrice + executionPrice * gasUsed` is `fee`

The math looks good to me

File: l2geth/core/rollup_fee_test.go [hunk 6]
```diff
@@ -23,9 +23,14 @@ func TestCalculateRollupFee(t *testing.T) {
 			data := make([]byte, 0, tt.dataLen)
 			fee := CalculateRollupFee(data, tt.gasUsed, big.NewInt(tt.dataPrice), big.NewInt(tt.executionPrice))
 
-			dataFee := uint64((RollupBaseTxSize + len(data)) * int(tt.dataPrice))
+			zeroes, ones := zeroesAndOnes(data)
+			zeroesCost := zeroes * 4
+			onesCost := (96 + ones) * 16
```
Comment (tynes) [inline l2geth/core/rollup_fee_test.go:28]: Should this use the constants as well? Or intentionally covering the fact that the constants may change

Comment (gakonst) [reply l2geth/core/rollup_fee_test.go:28]: Yeah this was intentional

**Review by tynes — COMMENTED** [review]
Looks good to me pending passing integration tests. Using the exact values in the integration tests is a double edged sword, especially if they were not calculated out of band

**Comment by gakonst** [comment]
I agree but IMO we should try pinning down certain values as regression tests. If we see this becoming a pain in the future (I don't think it will) we can consider reverting, but IMO net good now. They were calculated by `console.log`'ing the value before running the assertion. Ideally, we'd do all these with snapshot tests.

**Review by tynes — APPROVED** [review]
Passing integration tests, good work @gakonst 🎉

**Pull request merged**